### PR TITLE
First stab at adding new logging framework

### PR DESCRIPTION
**Pull request by bhariharan**

### CODE REVIEW

File: libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKFileLogger.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <CocoaLumberjack/DDFileLogger.h>
+
+@interface SFSDKFileLogger : DDFileLogger
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKFileLogger.h]: Extends the default `Lumberjack` file logger and adds some customizations.

File: libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogFileManager.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <CocoaLumberjack/DDFileLogger.h>
+
+@interface SFSDKLogFileManager : DDLogFileManagerDefault
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogFileManager.h]: Extends the default `Lumberjack` log file manager and adds some customizations (the ability to name log file, etc.).

File: libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <CocoaLumberjack/DDLog.h>
+#import "SFSDKFileLogger.h"
+
+@interface SFSDKLogger : NSObject
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: New logger interface. Basically the same feature set as `Android`.

File: libs/SalesforceAnalytics/SalesforceAnalyticsTests/SFSDKLoggerTests.m [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+static NSString * const kTestLogLine2 = @"This is test log line 2!";
+static NSString * const kTestLogLine3 = @"This is test log line 3!";
+static NSString * const kTestLogLine4 = @"This is test log line 4!";
+unsigned long long const kDefaultMaxFileSize = 1024 * 1024; // 1 MB.
+
+@interface SFSDKLoggerTests : XCTestCase
+
+@end
+
+@implementation SFSDKLoggerTests
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/SalesforceAnalyticsTests/SFSDKLoggerTests.m]: Tests are same as the `Android` implementation too.

File: libs/SalesforceSDKCore/SalesforceSDKCore/SalesforceSDKCore.h [hunk 5]
```diff
@@ -127,6 +127,7 @@
 #import <SalesforceSDKCore/SFSDKLoginHostListViewController.h>
 #import <SalesforceSDKCore/SFSDKLoginHostStorage.h>
 #import <SalesforceSDKCore/SFSDKResourceUtils.h>
+#import <SalesforceSDKCore/SFSDKSafeMutableDictionary.h>
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/SalesforceSDKCore.h]: This was missing from a previous PR.

File: libs/SalesforceSDKCore/SalesforceSDKCoreTests/SFSDKSafeMutableDictionaryTests.m [hunk 6]
```diff
@@ -50,14 +50,14 @@ - (void)tearDown {
 - (void)testConcurrentReadWrites {
     NSArray *overwriteValues = [self generateTestValues];
     XCTestExpectation *writeExpectation = [self expectationWithDescription:@"writeExpectation"];
-    XCTestExpectation *readExpactation = [self expectationWithDescription:@"readExpactation"];
+    XCTestExpectation *readExpectation = [self expectationWithDescription:@"readExpectation"];
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCoreTests/SFSDKSafeMutableDictionaryTests.m]: Fixed a minor typo.

**Comment by bhariharan** [comment]
Only outstanding work item is to add persistence for `maxFileSize` and `logLevel`.

File: libs/SalesforceAnalytics/SalesforceAnalyticsTests/SFSDKLoggerTests.m [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    XCTAssertEqual(newMaxSize, logger.fileLogger.maximumFileSize, @"Max size didn't match expected max size");
+}
+
+/**
+ * Test for flushing the log file.
+ */
+- (void)testFlushLogFile {
+    SFSDKLogger *logger = [SFSDKLogger sharedInstanceWithComponent:kTestComponent1];
+    XCTAssertEqualObjects(nil, [logger.fileLogger readFile], @"Log file should be empty");
+    [logger.logger log:NO message:[self messageForLogLine:kTestLogLine1]];
```
Comment (khawkins) [inline libs/SalesforceAnalytics/SalesforceAnalyticsTests/SFSDKLoggerTests.m]: It feels like we should decorate methods like `[DDLog  log:message:]` in the decorator class, if we're not going to go with a model of subclassing `DDLog`.  It seems cumbersome to have to call the underlying logger for what should be first-class operations on the decorating logging facility.

Comment (bhariharan) [inline libs/SalesforceAnalytics/SalesforceAnalyticsTests/SFSDKLoggerTests.m]: Yeah, that's in my PR that's up now.

**Comment by khawkins** [comment]
Minor comments.  Overall, looks fine to me.

**Pull request merged**